packages/cli/commands/project/cloneApp.js

```diff
@@ -0,0 +1,177 @@
+const {
+  uiLine,
+  uiCommandReference,
+  uiAccountDescription,
+} = require('../../lib/ui');
+const SpinniesManager = require('../../lib/ui/SpinniesManager');
+const {
+  logApiErrorInstance,
+  ApiErrorContext,
+} = require('../../lib/errorHandlers/apiErrors');
+const { EXIT_CODES } = require('../../lib/enums/exitCodes');
+const { isAppDeveloperAccount } = require('../../lib/accountTypes');
+const {
+  cloneApp,
+  checkCloneStatus,
+  downloadClonedProject,
+} = require('@hubspot/local-dev-lib/api/projects');
+const { getCwd, isValidPath } = require('@hubspot/local-dev-lib/path');
+const { logger } = require('@hubspot/local-dev-lib/logger');
+const { getAccountConfig } = require('@hubspot/local-dev-lib/config');
+const {
+  fetchPublicAppMetadata,
+} = require('@hubspot/local-dev-lib/api/appsDev');
+const { extractZipArchive } = require('@hubspot/local-dev-lib/archive');
+
+const i18nKey = 'commands.project.subcommands.cloneApp';
+
+exports.command = 'clone-app [--appId] [--location]';
+exports.describe = null; // uiBetaTag(i18n(`${i18nKey}.describe`), false);
+
+exports.handler = async options => {
+  await loadAndValidateOptions(options);
+
+  const accountId = getAccountId(options);
+  const accountConfig = getAccountConfig(accountId);
+  const accountName = uiAccountDescription(accountId);
+
+  trackCommandUsage('clone-app', {}, accountId);
+
+  if (!isAppDeveloperAccount(accountConfig)) {
+    uiLine();
+    logger.error(i18n(`${i18nKey}.errors.invalidAccountTypeTitle`));
+    logger.log(
+      i18n(`${i18nKey}.errors.invalidAccountTypeDescription`, {
+        useCommand: uiCommandReference('hs accounts use'),
+        authCommand: uiCommandReference('hs auth'),
+      })
+    );
+    uiLine();
+    process.exit(EXIT_CODES.ERROR);
+  }
+
+  let appId;
+  let appName;
+  let location;
+  try {
+    const appIdResponse =
+      'appId' in options
+        ? { appId: options.appId }
+        : await selectPublicAppPrompt({
+            accountId,
+            accountName,
+            options,
+            migrateApp: false,
+          });
+    appId = appIdResponse.appId;
+
+    const selectedApp = await fetchPublicAppMetadata(appId, accountId);
+    appName = selectedApp.name;
+
+    const locationResponse =
+      'location' in options
+        ? { location: options.location }
+        : await promptUser({
+            name: 'location',
+            message: i18n(`${i18nKey}.enterLocation`),
+            default: path.resolve(getCwd(), appName),
+            validate: input => {
+              if (!input) {
+                return i18n(`${i18nKey}.errors.locationRequired`);
+              }
+              if (fs.existsSync(input)) {
+                return i18n(`${i18nKey}.errors.invalidLocation`);
+              }
+              if (!isValidPath(input)) {
+                return i18n(`${i18nKey}.errors.invalidCharacters`);
+              }
+              return true;
+            },
+          });
```

Comment:
I know this isn't a huge prompt but might be worth moving to its own file due to the validation function being pretty lengthy. In this case, rather than using the `'location' in options` pattern, you could use the `when` field in `promptUser` to avoid running the prompt when location already exists (see `projectNamePrompt` for example.)

I would have suggested that for `selectPublicAppPrompt` too, but that prompt makes a request so I think it makes sense to skip calling it completely when we can.

Reply:
Addressed in https://github.com/HubSpot/hubspot-cli/pull/1085/commits/247b941649560ed561734257bca012ddf2d65f02